library/src/image.cxx

```diff
@@ -351,6 +353,145 @@ std::vector<unsigned char> image::saveBuffer(SaveFormat format) const
   }
 }
 
+//----------------------------------------------------------------------------
+std::string image::toTerminalText() const
+{
+  const int depth = this->getChannelCount();
+  if (this->getChannelType() != ChannelType::BYTE || depth < 3 || depth > 4)
+  {
+    throw std::invalid_argument("image must be byte RGB or RGBA");
+  }
+
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  const int width = dims[0];
+  const int height = dims[1];
+  const unsigned char* content = static_cast<unsigned char*>(this->getContent());
+
+  /* Function to retrieve pixels so we can return:
+    - transparent black values for out-of-bounds coords,
+    - opaque alpha value when the image has no alpha channel.
+    Rendering with half blocks means 1 line of text represents 2 rows of pixels
+    so we _will_ attempt to access a line past the bottom if the height is not even.
+  */
+  const auto getPixel = [&content, width, height, depth](int x, int y, unsigned char* alpha)
+  {
+    if (x >= 0 && x < width && y >= 0 && y < height)
+    {
+      const size_t i = depth * ((height - 1 - y) * width + x);
+      *alpha = depth > 3 ? content[i + 3] : (unsigned char)255;
+      return content[i + 0] << 16 | content[i + 1] << 8 | content[i + 2];
+    }
+    *alpha = (unsigned char)0;
+    return 0;
+  };
+
+  /* Functions to manipulate the terminal colors using escape sequences.
+    Keep track of the foreground and background states to avoid redundant sequences.
+  */
+  std::stringstream stream;
+  int currentFg = -1;
+  int currentBg = -1;
+  const auto setFg = [&stream, &currentFg](int rgb)
+  {
+    if (currentFg != rgb)
+    {
+      stream << "\033[38;2;" // set 24-bit foreground
+             << ((rgb >> 16) & 0xff) << ";" << ((rgb >> 8) & 0xff) << ";" << (rgb & 0xff) << "m";
+      currentFg = rgb;
+    }
+  };
+  const auto setBg = [&stream, &currentBg](int rgb)
+  {
+    if (currentBg != rgb)
+    {
+      stream << "\033[48;2;" // set 24-bit background
+             << ((rgb >> 16) & 0xff) << ";" << ((rgb >> 8) & 0xff) << ";" << (rgb & 0xff) << "m";
+      currentBg = rgb;
+    }
+  };
+  const auto reset = [&stream, &currentBg, &currentFg]()
+  {
+    if (currentBg > -1 || currentFg > -1)
+    {
+      stream << "\033[0m"; // reset all
+      currentBg = -1;
+      currentFg = -1;
+    }
+  };
+  const auto resetBg = [&stream, &currentBg]()
+  {
+    if (currentBg > -1)
+    {
+      stream << "\033[49m"; // reset background
+      currentBg = -1;
+    }
+  };
+
+  constexpr std::string_view EMPTY_BLOCK = " ";
+  // clang-format off
+  constexpr std::string_view TOP_BLOCK = u8"\u2580";    // U+2580
+  constexpr std::string_view BOTTOM_BLOCK = u8"\u2584"; // U+2584
+  constexpr std::string_view FULL_BLOCK = u8"\u2588";   // U+2588
+  // clang-format on
+  constexpr std::string_view EOL = "\n";
+  const unsigned char alphaCutoff = 127;
+
+  unsigned char a1, a2;
+  for (int y = 0; y < height; y += 2)
```

Comment:
What happens if the height is not even?

Reply:
`getPixel` returns transparent black for the non-existing pixels and the bottom row is drawn with only top half blocks with no background